docs/design/cronjob.md

```diff
@@ -0,0 +1,118 @@
+    // Optional deadline in seconds for starting the job if it misses scheduled
+    // time for any reason.  Missed jobs executions will be counted as failed ones.
+    // +optional
+    StartingDeadlineSeconds *int64 `json:"startingDeadlineSeconds,omitempty" protobuf:"varint,3,opt,name=startingDeadlineSeconds"`
+
+    // Specifies how to treat concurrent executions of a Job.
+    // Valid values are:
+    // - "Allow" (default): allows CronJobs to run concurrently;
+    // - "Forbid": forbids concurrent runs, skipping next run if previous run hasn't finished yet;
+    // - "Replace": cancels currently running job and replaces it with a new one
+    // +optional
+    ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy,omitempty" protobuf:"bytes,4,opt,name=concurrencyPolicy,casttype=ConcurrencyPolicy"`
+
+    // This flag tells the controller to suspend subsequent executions, it does
+    // not apply to already started executions.  Defaults to false.
+    // +optional
+    Suspend *bool `json:"suspend,omitempty" protobuf:"varint,5,opt,name=suspend"`
+    
+    // The number of successful finished jobs to retain.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    // Defaults to 3.
+    // +optional
+    SuccessfulJobsHistoryLimit *int32 `json:"successfulJobsHistoryLimit,omitempty" protobuf:"varint,6,opt,name=successfulJobsHistoryLimit"`
+
+    // The number of failed finished jobs to retain.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    // Defaults to 1.
+    // +optional
+    FailedJobsHistoryLimit *int32 `json:"failedJobsHistoryLimit,omitempty" protobuf:"varint,7,opt,name=failedJobsHistoryLimit"`
+}
+
+// ConcurrencyPolicy describes how the job will be handled.
+// Only one of the following concurrent policies may be specified.
+// If none of the following policies is specified, the default one
+// is AllowConcurrent.
+type ConcurrencyPolicy string
+
+const (
+	// AllowConcurrent allows CronJobs to run concurrently.
+	AllowConcurrent ConcurrencyPolicy = "Allow"
+
+	// ForbidConcurrent forbids concurrent runs, skipping next run if previous
+	// hasn't finished yet.
+	ForbidConcurrent ConcurrencyPolicy = "Forbid"
+
+	// ReplaceConcurrent cancels currently running job and replaces it with a new one.
+	ReplaceConcurrent ConcurrencyPolicy = "Replace"
+)
+
+// CronJobStatus represents the current state of a cron job.
+type CronJobStatus struct {
+	// A list of pointers to currently running jobs.
+	// +optional
+	Active []v1.ObjectReference `json:"active,omitempty" protobuf:"bytes,1,rep,name=active"`
+
+	// Information when was the last time the job was successfully scheduled.
+	// +optional
+	LastScheduleTime *metav1.Time `json:"lastScheduleTime,omitempty" protobuf:"bytes,4,opt,name=lastScheduleTime"`
+}
+
+```
+
+### CronJob Controller
+
+The `CronJobController` will run periodically, watching all `CronJob` resource as well as their batch job resources.
```

Comment:
what's the different between this implementation and spark-operator/tf-operator?

Reply:
There is no huge difference between them, the thing I can think of is kubernetes will perform the sync logic periodically.

Reply:
I'd suggest to reference spark/tf-operator scheduled job here.

Reply:
What do you mean here, replace the whole design/CronJob object within the design that in spark-operator(tf-operator doesn't have a Cron TFjob resource) or just the implementation detail? I personally prefer the later if this is you are asking.

Reply:
if I want to use vk for spark scheduled-application, what should I do?

Reply:
@k82cn AFAIK, when considering using vk for spark scheduled application, there is no big difference between these two approaches, since key point only lies at how we define the `JobSpec` or `JobTemplate` inside of CronJob, that is to say how can we define a volcano job to execute spark application,  these two approaches we mentioned only has difference on how to schedule the cron jobs.